coreax/solvers/recombination.py

```diff
@@ -0,0 +1,570 @@
+# © Crown Copyright GCHQ
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+r"""
+Recombination solvers.
+
+Take a dataset :math:`\{(x_i, w_i)\}_{i=1}^n`, where each node :math:`x_i \in \Omega`
+is paired with a weight :math:`w_i \in \mathbb{R} \ge 0`, and the sum of all weights
+is one, :math:`\sum_{i=1}^n w_i = 1` (a strict requirement for a probability measure).
+
+.. note::
+    Given any weighted dataset, we can use normalisation to satisfy the sum to
+    one condition, providing :math:`\sum_{i=1}^n w_i \neq 0`.
+
+Combined with :math:`m-1` test-functions :math:`\Phi^\prime = \{\phi_i\}_{i=1}^{m-1}`,
+where :math:`\phi_i \colon \Omega \to \mathbb{R}`, that parametrise a set of :math:`m`
+test-functions :math:`\Phi = \{x \mapsto 1\} \cup \Phi^\prime`, there exists a dataset
+push-forward measure :math:`\mu_n := \Phi_* \nu_n`.
+
+A recombination solver attempts to find a reduced measure (a coresubset)
+:math:`\hat{\mu}_{m^\prime}`, which is given as a basic-feasible solution (BFS) to the
+following linear-programming problem (with trivial objective)
+
+.. math::
+    \begin{align}
+        \mathbf{Y} \mathbf{\hat{w}} &= \text{CoM}(\mu_n),\\
+        \mathbf{\hat{w}} &\ge \mathbf{0},
```

Comment:
Probably worth defining \mathbf{0}

Reply:
I've probably used the wrong notation here. Should be `\mathbf{\hat{w}} &\ge 0`, where it is implicit that the greater than is elementwise. Do you think that would be clear?

Reply:
Yeah that would be better I think !